src/Symfony/Component/Messenger/Transport/Sender/SendersLocator.php

```diff
@@ -46,9 +46,9 @@ public function getSenders(Envelope $envelope, ?bool &$handle = false): iterable
         $seen = array();
 
         foreach (HandlersLocator::listTypes($envelope) as $type) {
-            foreach ($this->senders[$type] ?? array() as $sender) {
+            foreach ($this->senders[$type] ?? array() as $name => $sender) {
                 if (!\in_array($sender, $seen, true)) {
-                    yield $seen[] = $sender;
+                    yield \is_string($name) ? $name : \get_class($sender) => $seen[] = $sender;
```

Comment:
I think this logic should be moved  to the middleware: let's reduce the burden on SenderLocators implementations.